File: mnheme_ports/java/mnheme/Memory.java
// mnheme/java/mnheme/Memory.java
package mnheme;

import java.util.List;
import java.util.Map;

/**
 * Immutable snapshot of a memory record.
 * Using Java 16+ record for conciseness.
 */
public record Memory(
        String       memoryId,
        String       concept,
        String       feeling,
        String       mediaType,
        String       content,
        String       note,
        List<String> tags,
        String       timestamp,
        String       checksum
) {
    /** Convert to plain Map for JSON serialisation. */
    public Map<String, Object> toMap() {
        return Map.of(
            "memory_id",  memoryId,
            "concept",    concept,
            "feeling",    feeling,
            "media_type", mediaType,
            "content",    content,
            "note",       note,
            "tags",       tags,
            "timestamp",  timestamp,
            "checksum",   checksum
        );
    }

    @SuppressWarnings("unchecked")
    public static Memory fromMap(Map<String, Object> m) {
        return new Memory(
            (String)  m.getOrDefault("memory_id",  ""),
            (String)  m.getOrDefault("concept",    ""),
            (String)  m.getOrDefault("feeling",    ""),
            (String)  m.getOrDefault("media_type", "text"),
            (String)  m.getOrDefault("content",    ""),
            (String)  m.getOrDefault("note",       ""),
            m.containsKey("tags") ? (List<String>) m.get("tags") : List.of(),
            (String)  m.getOrDefault("timestamp",  ""),
            (String)  m.getOrDefault("checksum",   "")
        );
    }
}

File: mnheme_ports/kotlin/Memory.kt
// mnheme/kotlin/mnheme/Memory.kt
package mnheme

import org.json.JSONArray
import org.json.JSONObject

/**
 * Immutable snapshot of a stored memory.
 * data class gives free equals/hashCode/copy/toString.
 */
data class Memory(
    val memoryId:  String,
    val concept:   String,
    val feeling:   String,
    val mediaType: String,
    val content:   String,
    val note:      String,
    val tags:      List<String>,
    val timestamp: String,
    val checksum:  String,
) {
    fun toJson(): JSONObject = JSONObject().apply {
        put("memory_id",  memoryId)
        put("concept",    concept)
        put("feeling",    feeling)
        put("media_type", mediaType)
        put("content",    content)
        put("note",       note)
        put("tags",       JSONArray(tags))
        put("timestamp",  timestamp)
        put("checksum",   checksum)
    }

    companion object {
        fun fromJson(j: JSONObject): Memory {
            val tagsArr = j.optJSONArray("tags")
            val tags = if (tagsArr != null) {
                (0 until tagsArr.length()).map { tagsArr.optString(it) }
            } else emptyList()

            return Memory(
                memoryId  = j.optString("memory_id"),
                concept   = j.optString("concept"),
                feeling   = j.optString("feeling"),
                mediaType = j.optString("media_type", "text"),
                content   = j.optString("content"),
                note      = j.optString("note"),
                tags      = tags,
                timestamp = j.optString("timestamp"),
                checksum  = j.optString("checksum"),
            )
        }
    }
}
